core/src/main/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProvider.scala

```diff
@@ -282,42 +292,150 @@ class LocalEntityProvider(requestArguments: EntityRequestArguments,
       }
     }
 
-  override def queryEntities(entityType: String,
-                             query: EntityQuery,
-                             parentContext: RawlsRequestContext = requestArguments.ctx
-  ): Future[EntityQueryResponse] =
+  /**
+    * Returns the components needed to stream a EntityQueryResponse to an end user in response to the entityQuery API.
+    * This method returns fully materialized metadata (row counts, page size, etc) as EntityQueryResultMetadata, and
+    * also returns a streaming Source of Entity objects. We avoid materializing the full set of Entity objects for
+    * performance and memory reasons.
+    *
+    * @param entityType the type of entities to return in the result set
+    * @param query criteria for filtering and paginating the result set
+    * @param parentContext tracing context into which this method will add traces
+    * @return a tuple of 1) the fully materialized metadata, and 2) a streaming Source of Entity objects
+    */
+  override def queryEntitiesSource(entityType: String,
+                                   query: EntityQuery,
+                                   parentContext: RawlsRequestContext = requestArguments.ctx
+  ): Future[(EntityQueryResultMetadata, Source[Entity, _])] = {
+    // look for a columnFilter that specifies the primary key for this entityType;
+    // such a columnFilter means we are filtering by name and can greatly simplify the underlying query.
+    val nameFilter: Option[String] = query.columnFilter match {
+      case Some(colFilter)
+          if colFilter.attributeName == AttributeName.withDefaultNS(
+            entityType + Attributable.entityIdAttributeSuffix
+          ) =>
+        Option(colFilter.term)
+      case _ => None
+    }
+
+    parentContext.tracingSpan.foreach { span =>
+      span.putAttribute("isFilterByName", OpenCensusAttributeValue.booleanAttributeValue(nameFilter.isDefined))
+    }
+    // if filtering by name, retrieve that entity directly, else do the full query:
+    nameFilter match {
+      case Some(entityName) =>
+        dataSource.inTransaction { dataAccess =>
+          dataAccess.entityQuery.loadSingleEntityForPage(workspaceContext, entityType, entityName, query)
+        } map { case (unfilteredCount, filteredCount, entity) =>
+          val pageCount = if (entity.nonEmpty) 1 else 0
+          val entitySource = if (entity.nonEmpty) Source.single(entity.head) else Source.empty
+
+          val metadata: EntityQueryResultMetadata =
+            EntityQueryResultMetadata(unfilteredCount, filteredCount, pageCount)
+          (metadata, entitySource)
+        }
+
+      case _ =>
+        traceWithParent("loadEntityPage", parentContext) { childContext =>
```

Comment:
This tracing does not encompass streaming the results, it probably just encompasses up until the start of streaming. Probably not much to do about it right now, I am just calling it out.